win32ss/gdi/ntgdi/freetype.c

```diff
@@ -945,6 +1167,214 @@ SubstituteFontRecurse(LOGFONTW* pLogFont)
+        &hKey);
+    if (!NT_SUCCESS(Status))
+        return Status;
+
+    // Load the FontLink entry
+    cbData = sizeof(szzFontLink);
+    Status = RegQueryValue(hKey, pLF->lfFaceName, REG_MULTI_SZ, szzFontLink, &cbData);
+    if (!NT_SUCCESS(Status) && SubstituteFontRecurse(pLF)) // Retry with substituted name
+    {
+        cbData = sizeof(szzFontLink);
+        Status = RegQueryValue(hKey, pLF->lfFaceName, REG_MULTI_SZ, szzFontLink, &cbData);
+    }
+
+    ZwClose(hKey); // Close the registry key
+
+    RtlStringCchCopyW(pChain->szFaceName, _countof(pChain->szFaceName), pLF->lfFaceName);
+
+    ASSERT(SZZ_GetSize(s_szzDefFontLink) <= sizeof(szzFontLink));
+    ASSERT(SZZ_GetSize(s_szzDefFixedPitchFontLink) <= sizeof(szzFontLink));
+
+    if (!NT_SUCCESS(Status)) // Registry value not found
+    {
+        if (!(pLF->lfPitchAndFamily & FIXED_PITCH))
+            RtlCopyMemory(szzFontLink, s_szzDefFontLink, SZZ_GetSize(s_szzDefFontLink));
+        else
+            RtlCopyMemory(szzFontLink, s_szzDefFixedPitchFontLink,
+                          SZZ_GetSize(s_szzDefFixedPitchFontLink));
+    }
+
+    // Ensure double-NUL-terminated
+    szzFontLink[_countof(szzFontLink) - 1] = UNICODE_NULL;
+    szzFontLink[_countof(szzFontLink) - 2] = UNICODE_NULL;
+
+    FontLinkSize = SZZ_GetSize(szzFontLink);
+    pszzFontLink = ExAllocatePoolWithTag(PagedPool, FontLinkSize, TAG_FONT);
+    if (!pszzFontLink)
+        return STATUS_NO_MEMORY;
+    RtlCopyMemory(pszzFontLink, szzFontLink, FontLinkSize);
+    pChain->pszzFontLink = pszzFontLink;
+
+    return STATUS_SUCCESS;
+}
+
+static PFONTLINK
+FontLink_CreateOrGet(
+    _In_ const LOGFONTW *plfBase,
+    _In_ LPCWSTR pszLink)
+{
+    LPWSTR pch1, pch2;
+    LOGFONTW lf;
+    PFONTLINK pLink;
+
+    lf = *plfBase;
+    lf.lfCharSet = DEFAULT_CHARSET;
+
+    // pszLink: "<FontFileName>,<FaceName>[,...]"
+    pch1 = wcschr(pszLink, L',');
+    if (!pch1)
+    {
+        DPRINT1("%S\n", pszLink);
+        return NULL;
+    }
+    ++pch1;
+
+    pch2 = wcschr(pch1, L',');
+    if (pch2)
+        RtlStringCchCopyNW(lf.lfFaceName, _countof(lf.lfFaceName), pch1, pch2 - pch1);
+    else
+        RtlStringCchCopyW(lf.lfFaceName, _countof(lf.lfFaceName), pch1);
+
+    SubstituteFontRecurse(&lf);
+    DPRINT1("lfFaceName: %S\n", lf.lfFaceName);
+
+    pLink = ExAllocatePoolZero(PagedPool, sizeof(FONTLINK), TAG_FONT);
+    if (!pLink)
+    {
+        DPRINT1("!pLink\n");
+        return NULL;
+    }
+
+    pLink->LogFont = lf;
+
+    if (_wcsicmp(plfBase->lfFaceName, lf.lfFaceName) == 0)
+        pLink->bIgnoreThis = TRUE;
+
+    return pLink;
+}
+
+static NTSTATUS
+FontLink_Chain_Populate(
+    _Inout_ PFONTLINK_CHAIN pChain)
+{
+    NTSTATUS Status;
+    PFONTLINK pLink;
+    LOGFONTW lfBase;
+    PTEXTOBJ pTextObj = pChain->pBaseTextObj;
+    PFONTGDI pFontGDI;
+    PZZWSTR pszzFontLink;
+    WCHAR *pchLink;
```

Comment:
```suggestion
    PWSTR pchLink;
```
Could also be called `pszLink`